### ISSUE TITLE
win process spawn: race condition in inheriting handles

### ISSUE THREAD

**Issue by vtjnash** [libuv/libuv#1490]
If two threads were to simultaneously call `uv_spawn`, they might accidentally both inherit handles intended for the other process. Windows Vista introduced an API to help resolve this: "PROC_THREAD_ATTRIBUTE_HANDLE_LIST"
https://blogs.msdn.microsoft.com/oldnewthing/20111216-00/?p=8873

**Comment by bzoz** [comment]
I think all handles created by libuv are inheritable. IMHO  selecting which handles are to be passed to child process feels so Windows-specific, that users should call `CreateProcess` themselves.

**Comment by vtjnash** [comment]
I see refack self-assigned. You're also welcome to assign me. Passing `stdio` to a process is pretty basic functionality in a spawn API, and is already exposed by libuv. I've just recently noticed that there's a race condition in libuv due to this, but that Windows now provides an API to partially address it.

If you see any handles being created as inheritable, that sounds like a (separate) bug.

**Comment by bzoz** [comment]
Sorry, got confused. Handles are not inherited by default.

Yes, we should add this to libuv.

**Comment by egorpugin** [comment]
I was hit by this too.
Why inherit all handles in CreateProcessW() call?

**Comment by vtjnash** [comment]
Still might be nice to close this race condition.

**Comment by vtjnash** [comment]
No-stale

**Comment by santigimeno** [comment]
@vtjnash I've been looking a bit into this. Would implementing this mean that inheriting handles like [this](https://github.com/libuv/libuv/blob/5102b2c093681dae2a90ea2196f868de78ec9957/test/test-spawn.c#L1611-L1620) won't work anymore?

**Comment by vtjnash** [comment]
Correct. That is actually a bit of a gotcha here, since you can no longer accidentally leak an important handle–but also cannot intentionally leak them either, only explicitly.

Thus it may need to be an opt-in feature flag of uv_spawn, so that each application can decide whether they want security by default or compatibility behavior here.
